js/ui/paygUnlockDialog.js

```diff
@@ -0,0 +1,395 @@
+// -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-
+
+const Atk = imports.gi.Atk;
+const Clutter = imports.gi.Clutter;
+const Gio = imports.gi.Gio;
+const GLib = imports.gi.GLib;
+const GObject = imports.gi.GObject;
+const Lang = imports.lang;
+const Meta = imports.gi.Meta;
+const Pango = imports.gi.Pango;
+const Signals = imports.signals;
+const Shell = imports.gi.Shell;
+const St = imports.gi.St;
+
+const PaygManager = imports.misc.paygManager;
+
+const Animation = imports.ui.animation;
+const Main = imports.ui.main;
+const Monitor = imports.ui.monitor;
+const ShellEntry = imports.ui.shellEntry;
+const Tweener = imports.ui.tweener;
+
+// The timeout before going back automatically to the lock screen (in seconds)
+const IDLE_TIMEOUT = 2 * 60;
+
+const CODE_ENTRY_REQUIRED_LENGTH = 8;
+
+var SPINNER_ICON_SIZE = 16;
+var SPINNER_ANIMATION_DELAY = 1.0;
+var SPINNER_ANIMATION_TIME = 0.3;
+
+var MESSAGE_FADE_OUT_ANIMATION_TIME = 0.5;
+
+var UnlockStatus = {
+    NOT_VERIFYING: 0,
+    VERIFYING: 1,
+    FAILED: 2,
+    SUCCEEDED: 3
+};
+
+var PaygUnlockCodeEntry = new Lang.Class({
+    Name: 'PaygUnlockCodeEntry',
+    Extends: St.Entry,
+    Signals: { 'code-changed' : { param_types: [GObject.TYPE_STRING] } },
+
+    _init: function(params) {
+        this.parent({ style_class: 'unlock-dialog-payg-entry',
+                      reactive: true,
+                      can_focus: true,
+                      x_align: Clutter.ActorAlign.FILL });
+
+        this._code = '';
+        this.clutter_text.ellipsize = Pango.EllipsizeMode.NONE;
+        this.clutter_text.x_align = Clutter.ActorAlign.CENTER;
+
+        this.connect('button-press-event', this._onButtonPressEvent.bind(this));
+
+        this.clutter_text.connect('captured-event', this._onCapturedEvent.bind(this));
+        this.clutter_text.connect('text-changed', this._onTextChanged.bind(this));
+    },
+
+    _onCapturedEvent: function(textActor, event) {
+        if (event.type() != Clutter.EventType.KEY_PRESS)
+            return Clutter.EVENT_PROPAGATE;
+
+        let character = event.get_key_unicode();
+
+        // We only support printable characters.
```

Comment:
Question: What about hitting ‘enter’? Should that be handled explicitly here, or does Clutter handle that already in some other code path?